cups/man/cupsaddsmb.man
.\"
.\" "$Id: cupsaddsmb.man,v 1.1.1.10 2004/06/05 02:42:32 jlovell Exp $"
.\"
.\"   cupsaddsmb man page for the Common UNIX Printing System (CUPS).
.\"
.\"   Copyright 1997-2003 by Easy Software Products.
.\"
.\"   These coded instructions, statements, and computer programs are the
.\"   property of Easy Software Products and are protected by Federal
.\"   copyright law.  Distribution and use rights are outlined in the file
.\"   "LICENSE.txt" which should have been included with this file.  If this
.\"   file is missing or damaged please contact Easy Software Products
.\"   at:
.\"
.\"       Attn: CUPS Licensing Information
.\"       Easy Software Products
.\"       44141 Airport View Drive, Suite 204
.\"       Hollywood, Maryland 20636-3111 USA
.\"
.\"       Voice: (301) 373-9603
.\"       EMail: cups-info@cups.org
.\"         WWW: http://www.cups.org
.\"
.TH cupsaddsmb 8 "Common UNIX Printing System" "27 September 2002" "Easy Software Products"
.SH NAME
cupsaddsmb \- export printers to samba for windows clients
.SH SYNOPSIS
.B cupsaddsmb
[ -H
.I samba-server
] [ -U
.I samba-user
] [ -h
.I cups-server
] [ -v ] -a
.br
.B cupsaddsmb
[ -H
.I samba-server
] [ -U
.I samba-user
] [ -h
.I cups-server
] [ -v ] printer [ ... printer ]
.SH DESCRIPTION
\fIcupsaddsmb\fR exports printers to the SAMBA software (version
2.2.0 or higher) for use with Windows clients. Depending on the
SAMBA configuration, you may need to provide a password to
export the printers. This program requires the CUPS or Adobe PostScript
printer driver files described below.
.LP
The \fI-H\fR option specifies the SAMBA server which defaults
to the CUPS server.
.LP
The \fI-U\fR option specifies the SAMBA print admin username which defaults
to your current username.
.LP
The \fI-a\fR option exports all known printers. Otherwise only
the named printers are exported.
.LP
The \fI-h\fR option specifies a different CUPS server to use.
.LP
The \fI-v\fR option specifies that verbose information should be
shown and is useful for debugging SAMBA configuration problems.
.SH SAMBA CONFIGURATION
\fIcupsaddsmb\fR uses the new RPC-based printing support in
SAMBA 2.2.x to provide printer drivers and PPD files to Windows
client machines. In order to use this functionality, you must
first configure SAMBA (via the smb.conf file) to support
printing through CUPS and provide a printer driver download
share, as follows:
.nf

    [global]
	load printers = yes
	printing = cups
	printcap name = cups

    [printers]
	comment = All Printers
	path = /var/spool/samba
	browseable = no
	public = yes
	guest ok = yes
	writable = no
	printable = yes
	printer admin = root

    [print$]
	comment = Printer Drivers
	path = /etc/samba/drivers
	browseable = yes
	guest ok = no
	read only = yes
	write list = root
.fi
.LP
This configuration assumes a FHS-compliant installation of
SAMBA; adjust the [printers] and [print$] share paths
accordingly on your system as needed.

.SH CUPS POSTSCRIPT DRIVERS FOR WINDOWS
\fIcupsaddsmb\fR can use the CUPS PostScript printer drivers for
Windows, which are available for download from the CUPS web
site.
.LP
The CUPS printer drivers are preferred over the Adobe drivers
since they support page accounting, banner pages, and page
labels on all printers.  However, currently only Windows NT,
2000, and XP are supported by the CUPS drivers, so you will also
need to get the Adobe driver to support Windows 95, 98, and Me
clients.
.LP
Once you have extracted the driver files, create a "drivers"
directory in the CUPS data directory (usually /usr/share/cups)
and copy the CUPS files using lowercase filenames, as follows:
.nf

    [Windows NT, 2000, and XP]
    cups.hlp
    cupsdrvr.dll
    cupsui.dll
.fi

.SH ADOBE POSTSCRIPT DRIVERS FOR WINDOWS
\fIcupsaddsmb\fR can use the Adobe PostScript printer drivers for
Windows, which are available for download from the Adobe web
site (http://www.adobe.com).
.LP
The CUPS printer drivers are preferred over the Adobe drivers
since they support page accounting, banner pages, and page
labels on all printers.
.LP
Once you have extracted the driver files, create a "drivers"
directory in the CUPS data directory (usually /usr/share/cups)
and copy the Adobe files using UPPERCASE filenames, as follows:
.nf

    [Windows 95, 98, and Me]
    ADFONTS.MFM
    ADOBEPS4.DRV
    ADOBEPS4.HLP
    DEFPRTR2.PPD
    ICONLIB.DLL
    PSMON.DLL

    [Windows NT, 2000, and XP]
    ADOBEPS5.DLL
    ADOBEPSU.DLL
    ADOBEPSU.HLP
.fi
.LP
You can download and copy either set of driver files
independently, and use the CUPS drivers for Windows NT, 2000,
and XP while using the Adobe drivers for the Windows 95, 98, and
Me clients.
.SH SEE ALSO
CUPS Software Administrators Manual,
http://localhost:631/documentation.html
.SH COPYRIGHT
Copyright 1993-2003 by Easy Software Products, All Rights Reserved.
.\"
.\" End of "$Id: cupsaddsmb.man,v 1.1.1.10 2004/06/05 02:42:32 jlovell Exp $".
.\"


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

cups/systemv/cupsaddsmb.c
/*
 * "$Id: cupsaddsmb.c,v 1.1.1.14 2004/06/05 02:42:35 jlovell Exp $"
 *
 *   "cupsaddsmb" command for the Common UNIX Printing System (CUPS).
 *
 *   Copyright 2001-2004 by Easy Software Products.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Easy Software Products and are protected by Federal
 *   copyright law.  Distribution and use rights are outlined in the file
 *   "LICENSE.txt" which should have been included with this file.  If this
 *   file is missing or damaged please contact Easy Software Products
 *   at:
 *
 *       Attn: CUPS Licensing Information
 *       Easy Software Products
 *       44141 Airport View Drive, Suite 204
 *       Hollywood, Maryland 20636-3111 USA
 *
 *       Voice: (301) 373-9603
 *       EMail: cups-info@cups.org
 *         WWW: http://www.cups.org
 *
 * Contents:
 *
 *   main()             - Export printers on the command-line.
 *   do_samba_command() - Do a SAMBA command, asking for a password as needed.
 *   export_dest()      - Export a destination to SAMBA.
 *   usage()            - Show program usage and exit...
 */

/*
 * Include necessary headers...
 */

#include <stdio.h>
#include <stdlib.h>
#include <cups/cups.h>
#include <cups/language.h>
#include <cups/string.h>
#include <errno.h>


/*
 * Local globals...
 */

int		Verbosity = 0;
const char	*SAMBAUser,
		*SAMBAServer;


/*
 * Local functions...
 */

int	do_samba_command(const char *, const char *);
int	export_dest(const char *);
void	usage();


/*
 * 'main()' - Export printers on the command-line.
 */

int			/* O - Exit status */
main(int  argc,		/* I - Number of command-line arguments */
     char *argv[])	/* I - Command-line arguments */
{
  int	i, j;		/* Looping vars */
  int	status;		/* Status from export_dest() */
  int	export_all;	/* Export all printers? */
  int	num_printers;	/* Number of printers */
  char	**printers;	/* Printers */


 /*
  * Parse command-line arguments...
  */

  export_all = 0;

  SAMBAUser   = cupsUser();
  SAMBAServer = NULL;

  for (i = 1; i < argc; i ++)
    if (strcmp(argv[i], "-a") == 0)
      export_all = 1;
    else if (strcmp(argv[i], "-U") == 0)
    {
      i ++;
      if (i >= argc)
        usage();

      SAMBAUser = argv[i];
    }
    else if (strcmp(argv[i], "-H") == 0)
    {
      i ++;
      if (i >= argc)
        usage();

      SAMBAServer = argv[i];
    }
    else if (strcmp(argv[i], "-h") == 0)
    {
      i ++;
      if (i >= argc)
        usage();

      cupsSetServer(argv[i]);
    }
    else if (strcmp(argv[i], "-v") == 0)
      Verbosity = 1;
    else if (argv[i][0] != '-')
    {
      if (SAMBAServer == NULL)
	SAMBAServer = cupsServer();

      if ((status = export_dest(argv[i])) != 0)
	return (status);
    }
    else
      usage();

 /*
  * See if the user specified "-a"...
  */

  if (export_all)
  {
   /*
    * Export all printers...
    */

    if (SAMBAServer == NULL)
      SAMBAServer = cupsServer();

    num_printers = cupsGetPrinters(&printers);

    for (j = 0, status = 0; j < num_printers; j ++)
      if ((status = export_dest(printers[j])) != 0)
	break;

    for (j = 0; j < num_printers; j ++)
      free(printers[j]);

    if (num_printers)
      free(printers);

    if (status)
      return (status);
  }

  return (0);
}


/*
 * 'do_samba_command()' - Do a SAMBA command, asking for
 *                        a password as needed.
 */

int					/* O - Status of command */
do_samba_command(const char *command,	/* I - Command to run */
                 const char *subcmd)	/* I - Sub-command */
{
  int		status;			/* Status of command */
  char		temp[4096];		/* Command/prompt string */
  static const char *p = NULL;		/* Password data */


  for (status = 1;;)
  {
    if (p == NULL)
    {
      snprintf(temp, sizeof(temp),
               "Password for %s required to access %s via SAMBA: ",
	       SAMBAUser, SAMBAServer);

      if ((p = cupsGetPassword(temp)) == NULL)
	break;
    }

    snprintf(temp, sizeof(temp), "%s -N -U\'%s%%%s\' -c \'%s\'",
             command, SAMBAUser, p, subcmd);

    if (Verbosity)
      printf("Running command: %s\n", temp);
    else
      strlcat(temp, " </dev/null >/dev/null 2>/dev/null", sizeof(temp));

    if ((status = system(temp)) != 0)
    {
      if (Verbosity)
        puts("");

      if (p[0])
        p = NULL;
      else
        break;
    }
    else
    {
      if (Verbosity)
        puts("");

      break;
    }
  }

  return (status);
}


/*
 * 'export_dest()' - Export a destination to SAMBA.
 */

int					/* O - 0 on success, non-zero on error */
export_dest(const char *dest)		/* I - Destination to export */
{
  int			i;		/* Looping var */
  int			status;		/* Status of smbclient/rpcclient commands */
  const char		*ppdfile;	/* PPD file for printer drivers */
  char			command[1024],	/* Command to run */
			subcmd[1024];	/* Sub-command */
  const char		*datadir;	/* CUPS_DATADIR */
  FILE			*fp;		/* PPD file */
  http_t		*http;		/* Connection to server */
  cups_lang_t		*language;	/* Default language */
  ipp_t			*request,	/* IPP request */
			*response;	/* IPP response */
  ipp_attribute_t	*attr;		/* IPP attribute */
  static const char	*pattrs[] =	/* Printer attributes we want */
			{
			  "job-sheets-supported",
			  "job-sheets-default",
			  "job-hold-until-supported",
			  "job-hold-until-default"
			};


  if ((datadir = getenv("CUPS_DATADIR")) == NULL)
    datadir = CUPS_DATADIR;

 /*
  * Get the PPD file...
  */

  if ((ppdfile = cupsGetPPD(dest)) == NULL)
  {
    fprintf(stderr, "Warning: No PPD file for printer \"%s\" - skipping!\n", dest);
    return (0);
  }

 /*
  * Append the supported banner pages to the PPD file...
  */

  if ((http = httpConnectEncrypt(cupsServer(), ippPort(), cupsEncryption())) == NULL)
  {
    fprintf(stderr, "Unable to connect to server \"%s\" for %s - %s\n",
            cupsServer(), dest, strerror(errno));
    unlink(ppdfile);
    return (1);
  }

  request = ippNew();
  request->request.op.operation_id = IPP_GET_PRINTER_ATTRIBUTES;
  request->request.op.request_id   = 1;

  language = cupsLangDefault();

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
               "attributes-charset", NULL, cupsLangEncoding(language));

  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
               "attributes-natural-language", NULL, language->language);

  snprintf(command, sizeof(command), "ipp://localhost/printers/%s", dest);
  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
               "printer-uri", NULL, command);

  ippAddStrings(request, IPP_TAG_OPERATION, IPP_TAG_KEYWORD,
                "requested-attributes", sizeof(pattrs) / sizeof(pattrs[0]),
		NULL, pattrs);

 /*
  * Do the request and get back a response...
  */

  if ((response = cupsDoRequest(http, request, "/")) != NULL)
  {
    if (response->request.status.status_code > IPP_OK_CONFLICT)
    {
      fprintf(stderr, "ERROR: get-printer-attributes failed for %s: %s\n",
              dest, ippErrorString(response->request.status.status_code));
      ippDelete(response);
      unlink(ppdfile);
      return (2);
    }
  }
  else
  {
    fprintf(stderr, "ERROR: get-printer-attributes failed for %s: %s\n",
            dest, ippErrorString(cupsLastError()));
    unlink(ppdfile);
    return (2);
  }

 /*
  * Append the banner attributes to the end of the PPD file...
  */

  if ((fp = fopen(ppdfile, "a")) == NULL)
  {
    fprintf(stderr, "ERROR: Unable to append banner attributes to PPD file for %s - %s\n",
            dest, strerror(errno));
    ippDelete(response);
    unlink(ppdfile);
    return (3);
  }

  if ((attr = ippFindAttribute(response, "job-sheets-supported", IPP_TAG_NAME)) != NULL)
  {
    fprintf(fp, "*cupsJobSheetsSupported: \"%s", attr->values[0].string.text);

    for (i = 1; i < attr->num_values; i ++)
      fprintf(fp, ",%s", attr->values[i].string.text);

    fputs("\"\n", fp);
  }

  if ((attr = ippFindAttribute(response, "job-sheets-default", IPP_TAG_NAME)) != NULL)
  {
    fprintf(fp, "*cupsJobSheetsDefault: \"%s", attr->values[0].string.text);

    if (attr->num_values > 1)
      fprintf(fp, ",%s", attr->values[1].string.text);

    fputs("\"\n", fp);
  }

  if ((attr = ippFindAttribute(response, "job-hold-until-supported",
                               IPP_TAG_KEYWORD)) != NULL)
  {
    fprintf(fp, "*cupsJobHoldUntilSupported: \"%s", attr->values[0].string.text);

    for (i = 1; i < attr->num_values; i ++)
      fprintf(fp, ",%s", attr->values[i].string.text);

    fputs("\"\n", fp);
  }

  if ((attr = ippFindAttribute(response, "job-hold-until-default", IPP_TAG_KEYWORD)) != NULL)
    fprintf(fp, "*cupsJobHoldUntilDefault: \"%s\"\n", attr->values[0].string.text);

  fclose(fp);

  ippDelete(response);
  cupsLangFree(language);
  httpClose(http);

 /*
  * See which drivers are available - old CUPS, new CUPS, old Adobe, or
  * new Adobe?
  */

  snprintf(command, sizeof(command), "%s/drivers/cupsdrv5.dll", datadir);
  if (access(command, 0) == 0)
  {
   /*
    * Do the smbclient commands needed for the new CUPS WinNT drivers...
    */

    snprintf(command, sizeof(command), "smbclient //%s/print\\$", SAMBAServer);

    snprintf(subcmd, sizeof(subcmd),
             "mkdir W32X86;"
	     "put %s W32X86/%s.ppd;"
	     "put %s/drivers/cupsdrv5.dll W32X86/cupsdrv5.dll;"
	     "put %s/drivers/cupsui5.dll W32X86/cupsui5.dll;"
	     "put %s/drivers/cups5.hlp W32X86/cups5.hlp",
	     ppdfile, dest, datadir, datadir, datadir);

    if ((status = do_samba_command(command, subcmd)) != 0)
    {
      fprintf(stderr, "ERROR: Unable to copy Windows printer driver files (%d)!\n",
              status);
      unlink(ppdfile);
      return (4);
    }

   /*
    * Do the rpcclient commands needed for the CUPS WinNT drivers...
    */

    snprintf(subcmd, sizeof(subcmd),
             "adddriver \"Windows NT x86\" \"%s:cupsdrv5.dll:%s.ppd:cupsui5.dll:cups5.hlp:NULL:RAW:NULL\"",
	     dest, dest);

    snprintf(command, sizeof(command), "rpcclient %s", SAMBAServer);

    if ((status = do_samba_command(command, subcmd)) != 0)
    {
      fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
              status);
      unlink(ppdfile);
      return (5);
    }
  }
  else
  {
    snprintf(command, sizeof(command), "%s/drivers/cupsdrvr.dll", datadir);
    if (access(command, 0) == 0)
    {
     /*
      * Do the smbclient commands needed for the old CUPS WinNT drivers...
      */

      snprintf(command, sizeof(command), "smbclient //%s/print\\$", SAMBAServer);

      snprintf(subcmd, sizeof(subcmd),
               "mkdir W32X86;"
	       "put %s W32X86/%s.ppd;"
	       "put %s/drivers/cupsdrvr.dll W32X86/cupsdrvr.dll;"
	       "put %s/drivers/cupsui.dll W32X86/cupsui.dll;"
	       "put %s/drivers/cups.hlp W32X86/cups.hlp",
	       ppdfile, dest, datadir, datadir, datadir);

      if ((status = do_samba_command(command, subcmd)) != 0)
      {
	fprintf(stderr, "ERROR: Unable to copy Windows printer driver files (%d)!\n",
        	status);
	unlink(ppdfile);
	return (4);
      }

     /*
      * Do the rpcclient commands needed for the CUPS WinNT drivers...
      */

      snprintf(subcmd, sizeof(subcmd),
               "adddriver \"Windows NT x86\" \"%s:cupsdrvr.dll:%s.ppd:cupsui.dll:cups.hlp:NULL:RAW:NULL\"",
	       dest, dest);

      snprintf(command, sizeof(command), "rpcclient %s", SAMBAServer);

      if ((status = do_samba_command(command, subcmd)) != 0)
      {
	fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
        	status);
	unlink(ppdfile);
	return (5);
      }
    }
    else
    {
     /*
      * Do the smbclient commands needed for the old Adobe WinNT drivers...
      */

      snprintf(command, sizeof(command), "smbclient //%s/print\\$", SAMBAServer);

      snprintf(subcmd, sizeof(subcmd),
               "mkdir W32X86;"
	       "put %s W32X86/%s.PPD;"
	       "put %s/drivers/ADOBEPS5.DLL W32X86/ADOBEPS5.DLL;"
	       "put %s/drivers/ADOBEPSU.DLL W32X86/ADOBEPSU.DLL;"
	       "put %s/drivers/ADOBEPSU.HLP W32X86/ADOBEPSU.HLP",
	       ppdfile, dest, datadir, datadir, datadir);

      if ((status = do_samba_command(command, subcmd)) != 0)
      {
	fprintf(stderr, "ERROR: Unable to copy Windows printer driver files (%d)!\n",
        	status);
	unlink(ppdfile);
	return (4);
      }

     /*
      * Do the rpcclient commands needed for the Adobe WinNT drivers...
      */

      snprintf(subcmd, sizeof(subcmd),
               "adddriver \"Windows NT x86\" \"%s:ADOBEPS5.DLL:%s.PPD:ADOBEPSU.DLL:ADOBEPSU.HLP:NULL:RAW:NULL\"",
	       dest, dest);

      snprintf(command, sizeof(command), "rpcclient %s", SAMBAServer);

      if ((status = do_samba_command(command, subcmd)) != 0)
      {
	fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
        	status);
	unlink(ppdfile);
	return (5);
      }
    }
  }

  snprintf(command, sizeof(command), "%s/drivers/ADOBEPS4.DRV", datadir);
  if (access(command, 0) == 0)
  {
   /*
    * Do the smbclient commands needed for the Adobe Win9x drivers...
    */

    snprintf(command, sizeof(command), "smbclient //%s/print\\$", SAMBAServer);

    snprintf(subcmd, sizeof(subcmd),
             "mkdir WIN40;"
	     "put %s WIN40/%s.PPD;"
	     "put %s/drivers/ADFONTS.MFM WIN40/ADFONTS.MFM;"
	     "put %s/drivers/ADOBEPS4.DRV WIN40/ADOBEPS4.DRV;"
	     "put %s/drivers/ADOBEPS4.HLP WIN40/ADOBEPS4.HLP;"
	     "put %s/drivers/DEFPRTR2.PPD WIN40/DEFPRTR2.PPD;"
	     "put %s/drivers/ICONLIB.DLL WIN40/ICONLIB.DLL;"
	     "put %s/drivers/PSMON.DLL WIN40/PSMON.DLL;",
	     ppdfile, dest, datadir, datadir, datadir,
	     datadir, datadir, datadir);

    if ((status = do_samba_command(command, subcmd)) != 0)
    {
      fprintf(stderr, "ERROR: Unable to copy Windows printer driver files (%d)!\n",
              status);
      unlink(ppdfile);
      return (6);
    }

   /*
    * Do the rpcclient commands needed for the Adobe Win9x drivers...
    */

    snprintf(command, sizeof(command), "rpcclient %s", SAMBAServer);

    snprintf(subcmd, sizeof(subcmd),
	     "adddriver \"Windows 4.0\" \"%s:ADOBEPS4.DRV:%s.PPD:NULL:ADOBEPS4.HLP:PSMON.DLL:RAW:"
	     "ADOBEPS4.DRV,%s.PPD,ADOBEPS4.HLP,PSMON.DLL,ADFONTS.MFM,DEFPRTR2.PPD,ICONLIB.DLL\"",
	     dest, dest, dest);

    if ((status = do_samba_command(command, subcmd)) != 0)
    {
      fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
              status);
      unlink(ppdfile);
      return (7);
    }
  }

  unlink(ppdfile);

 /*
  * Finally, associate the drivers we just added with the queue...
  */

  snprintf(command, sizeof(command), "rpcclient %s", SAMBAServer);

  snprintf(subcmd, sizeof(subcmd), "setdriver %s %s", dest, dest);

  if ((status = do_samba_command(command, subcmd)) != 0)
  {
    fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
            status);
    return (8);
  }

  return (0);
}


/*
 * 'usage()' - Show program usage and exit...
 */

void
usage()
{
  puts("Usage: cupsaddsmb [options] printer1 ... printerN");
  puts("       cupsaddsmb [options] -a");
  puts("");
  puts("Options:");
  puts("  -H samba-server  Use the named SAMBA server");
  puts("  -U samba-user    Authenticate using the named SAMBA user");
  puts("  -a               Export all printers");
  puts("  -h cups-server   Use the named CUPS server");
  puts("  -v               Be verbose (show commands)");
  exit(1);
}


/*
 * End of "$Id: cupsaddsmb.c,v 1.1.1.14 2004/06/05 02:42:35 jlovell Exp $".
 */
